go-controller/pkg/factory/factory.go

```diff
@@ -56,6 +79,97 @@ func (i *informer) forEachHandler(obj interface{}, f func(h *Handler)) {
 	}
 }
 
+func (i *informer) processEvents(eq *eventQueue, namespacedName ktypes.NamespacedName) {
+	for {
+		select {
+		case event, ok := <-eq.queue:
+			if !ok {
+				return
+			}
+
+			switch event.kind {
+			case addEvent:
+				i.runAddHandlers(event.obj)
+			case updateEvent:
+				i.runUpdateHandlers(event.oldObj, event.obj)
+			case deleteEvent:
+				i.runDeleteHandlers(event.obj)
+			}
+			// If there are no more events left in the queue,
+			// drop this object's event queue from the informer's
+			// object map
+			if func() bool {
+				i.Lock()
+				defer i.Unlock()
+				eq.refcount--
+				if eq.refcount == 0 {
+					close(eq.queue)
+					delete(i.objectQueues, namespacedName)
+				}
+				return eq.refcount == 0
+			}() {
+				return
+			}
+		case <-eq.stopChan:
+			close(eq.queue)
+			return
+		}
+	}
+}
+
+func (i *informer) addToQueue(oldObj, obj interface{}, kind eventKind) {
+	meta, err := getObjectMeta(i.oType, obj)
+	if err != nil {
+		logrus.Errorf("object has no meta: %v", err)
+		return
+	}
+	namespacedName := ktypes.NamespacedName{
+		Namespace: meta.Namespace,
+		Name:      meta.Name,
+	}
+
+	i.Lock()
+	defer i.Unlock()
+
+	// Spawn a separate event processor for every object; that
+	// processor will serialize all events for the object but allow
+	// events for different objects to proceed in parallel
+	eq, ok := i.objectQueues[namespacedName]
+	if !ok {
+		eq = &eventQueue{
+			queue:    make(chan *event, 6),
```

Comment:
Justify `6`. (I was reading something recently suggesting that all channel sizes other than `0` or `1` are a code smell...)

Reply:
@danwinship I guess we can do 1... I was leaving some room for at least an Add + Update + Delete or so. No particular reason for 6 other than "it's greater than 0".

Reply:
> No particular reason for 6 other than "it's greater than 0".

So I guess in more detail, I was asking, is there any chance that the pushing thread is going to be pushing updates faster than the popping thread can process them (to which I think the answer is "yes"), and if so, does it matter if the pushing thread gets blocked in that case (which I'm not sure about), and if so, then what's the most it could plausibly lag? If it doesn't matter if the pushing thread gets blocked then you can just go with 0/unspecified and make it synchronous. If it does matter if the pushing thread gets blocked, then you need a channel size that's guaranteed to not fill up. (I guess the third case is, if it's annoying but non-fatal if the pushing thread blocks, then you want a capacity that *usually* won't overflow.)

Reply:
@danwinship how about now...